### PR TITLE
ci: Add support for `max_inflight_requests` parameter to prevent unbounded memory growth in ensemble models

### PR DESCRIPTION
**Pull request by pskiran1**
#### What does the PR do?

This PR adds testing for a new parameter `max_inflight_requests`.
Added `ensemble_backpressure_test.py` with custom decoupled producer and slow consumer models to validate the new feature.

#### Checklist
- [x] PR title reflects the change and is of format `<commit_type>: <Title>`
- [x] Changes are described in the pull request.
- [ ] Related issues are referenced.
- [x] Populated [github labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) field
- [x] Added [test plan](#test-plan) and verified test passes.
- [ ] Verified that the PR passes existing CI.
- [x] Verified copyright is correct on all changed files.
- [x] Added _succinct_ git squash message before merging [ref](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
- [ ] All template sections are filled out.
- [ ] Optional: Additional screenshots for behavior/output changes with before/after.

#### Commit Type:
Check the [conventional commit type](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type)
box here and add the label to the github PR.
- [ ] build
- [x] ci
- [ ] docs
- [ ] feat
- [ ] fix
- [ ] perf
- [ ] refactor
- [ ] revert
- [ ] style
- [ ] test

#### Related PRs:
https://github.com/triton-inference-server/core/pull/455
https://github.com/triton-inference-server/common/pull/141

#### Test plan:

- CI Pipeline ID: 37719578

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)
- closes GitHub issue: #xxx

### CODE REVIEW

File: qa/L0_simple_ensemble/backpressure_test_models/ensemble_disabled_max_inflight_requests/config.pbtxt [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# Copyright 2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (yinggeh) [inline qa/L0_simple_ensemble/backpressure_test_models/ensemble_disabled_max_inflight_requests/config.pbtxt:1]: You can reuse `config.pbtxt` from `ensemble_disabled_max_inflight_responses` and add the new parameter in test.sh.

File: qa/L0_simple_ensemble/test.sh [hunk 2]
```diff
@@ -146,6 +148,107 @@ set -e
 kill $SERVER_PID
 wait $SERVER_PID
 
+# Test ensemble backpressure feature (max_ensemble_inflight_responses parameter)
```
Comment (yinggeh) [inline qa/L0_simple_ensemble/test.sh]: Remove `/models` and make model repository here. It's easier for debugging and can speed up model loading time for other tests.

File: qa/L0_simple_ensemble/models/slow_consumer/1/model.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import time
+
+import triton_python_backend_utils as pb_utils
+
+
+class TritonPythonModel:
```
Comment (yinggeh) [inline qa/L0_simple_ensemble/models/slow_consumer/1/model.py:32]: Copy from `qa/python_models/ground_truth` which does the same thing.

File: qa/L0_simple_ensemble/test.sh (hunk 2 above)
Comment (pskiran1) [reply qa/L0_simple_ensemble/test.sh]: Applied the above change to have individual model repositories for different groups of test cases.

File: docs/user_guide/decoupled_models.md [hunk 4]
```diff
@@ -95,7 +95,13 @@ your application should be cognizant that the callback function you registered w
 `TRITONSERVER_InferenceRequestSetResponseCallback` can be invoked any number of times,
 each time with a new response. You can take a look at [grpc_server.cc](https://github.com/triton-inference-server/server/blob/main/src/grpc/grpc_server.cc)
 
-### Knowing When a Decoupled Inference Request is Complete
+### Using Decoupled Models in Ensembles
```
Comment (yinggeh) [inline docs/user_guide/decoupled_models.md:98]: Is it only happening in decoupled model, or any models with big processing speed difference?

File: docs/user_guide/ensemble_models.md [hunk 5]
```diff
@@ -183,6 +183,66 @@ performance, you can use
 [Model Analyzer](https://github.com/triton-inference-server/model_analyzer)
 to find the optimal model configurations.
 
+## Managing Memory Usage in Ensembles with Decoupled Models
+
+An *inflight response* is an intermediate output generated by an upstream model and held in memory until it is consumed by a downstream model within an ensemble pipeline. When an ensemble pipeline contains [decoupled models](decoupled_models.md) that produce responses faster than downstream models can process them, inflight responses accumulate internally and may cause unbounded memory growth. This commonly occurs in data preprocessing pipelines where a fast decoupled model (such as DALI, which efficiently streams and preprocesses data) feeds into a slower inference model (such as ONNX Runtime or TensorRT, which are compute-intensive and operate at a lower throughput).
+
+Consider an ensemble with two steps:
```
Comment (yinggeh) [inline docs/user_guide/ensemble_models.md]:
```suggestion
Consider an example ensemble model with two steps:
```

File: docs/user_guide/ensemble_models.md [hunk 6]
```diff
@@ -183,6 +183,66 @@ performance, you can use
...
+1. **DALI preprocessor** (decoupled): Produces 100 preprocessed images/sec
+2. **ONNX inference model**: Consumes 10 images/sec
+
+Here, the DALI model produces responses 10× faster than the ONNX model can process them. Without backpressure, these intermediate tensors accumulate in memory, eventually leading to out-of-memory errors.
+
+The `max_inflight_responses` field in the ensemble configuration limits the number of concurrent inflight responses between ensemble steps per request.
+When this limit is reached, faster upstream models are paused (blocked) until downstream models finish processing, effectively preventing unbounded memory growth.
```
Comment (yinggeh) [inline docs/user_guide/ensemble_models.md]: I have a concern of getting a deadlock here. How do you make sure the slower model always gets the chance to execute instead of all `max_inflight_responses` are blocked at the faster producer?

File: docs/user_guide/ensemble_models.md [hunk 7]
```diff
@@ -183,6 +183,66 @@ performance, you can use
...
+
+**Configuration:**
+* **`max_inflight_responses: 16`**: For each ensemble request (not globally), at most 16 responses from `dali_preprocess`
+  can wait for `onnx_inference` to process. Once this per-request limit is reached, `dali_preprocess` is blocked until the downstream step completes a response.
+* **Default (`0`)**: No limit - allows unlimited inflight responses (original behavior).
+
+### When to Use This Feature
+
+Use `max_inflight_responses` when your ensemble includes:
+* **Decoupled models** that produce multiple responses per request
```
Comment (yinggeh) [inline docs/user_guide/ensemble_models.md]: Same here

File: docs/user_guide/ensemble_models.md [hunk 8]
```diff
@@ -183,6 +183,66 @@ performance, you can use
...
+      input_map { key: "INPUT", value: "preprocessed" }
+      output_map { key: "OUTPUT", value: "RESULT" }
+    }
+  ]
+}
+```
+
+**Configuration:**
+* **`max_inflight_responses: 16`**: For each ensemble request (not globally), at most 16 responses from `dali_preprocess`
+  can wait for `onnx_inference` to process. Once this per-request limit is reached, `dali_preprocess` is blocked until the downstream step completes a response.
```
Comment (yinggeh) [inline docs/user_guide/ensemble_models.md]:
```suggestion
  can wait for `onnx_inference` to process. Once this per-step limit is reached, `dali_preprocess` is blocked until the downstream step completes a response.
```

File: docs/user_guide/decoupled_models.md [hunk 9]
```diff
@@ -95,7 +95,13 @@ your application should be cognizant that the callback function you registered w
...
+
+When using decoupled models within an [ensemble](ensemble_models.md), you may encounter unbounded memory growth if a decoupled model produces responses faster than downstream models can consume them. To address this, Triton provides the `max_inflight_responses` configuration field, which limits the number of concurrent inflight responses between ensemble steps.
```
Comment (yinggeh) [inline docs/user_guide/decoupled_models.md]:
```suggestion
When using decoupled models within an [ensemble](ensemble_models.md), you may encounter unbounded memory growth if a decoupled model produces responses faster than downstream models can consume them. To address this, Triton provides the `max_inflight_responses` configuration field, which limits the number of concurrent inflight responses at each ensemble step.
```

File: docs/user_guide/decoupled_models.md (hunk 4 above)
Comment (pskiran1) [reply docs/user_guide/decoupled_models.md:98]: Based on my understanding, since the normal model step will have only one response, a slow processing step will automatically block the request. In this case, memory usage will increase at a normal rate (always only one response per step accumulates in memory, unlike in a decoupled model) and may not need any additional backpressure at the step level. To effectively manage overall memory usage, a `rate limiter` could be sufficient.

Comment (yinggeh) [reply docs/user_guide/decoupled_models.md:98]: Maybe I am thinking a different senario or I am misunderstanding. Consider a simple two-step ensemble and first step processes at a much faster speed. The client sends request at a constant rate (e.g. one request per millisecond). Based on https://github.com/triton-inference-server/core/blob/b354d4dc13c5855b50a36eeec0d4d3aa443a01f3/src/ensemble_scheduler/ensemble_scheduler.cc#L1422, `ScheduleSteps` is non-blocking and the queue size increases in an constant rate.

In your change, since `ScheduleSteps` is called in `EnsembleScheduler::Enqueue` (same thread of http/grpc handler thread) and `ResponseComplete`, in the above example it will also blocks grpc/http handler and limit the total memory growth.

File: qa/L0_simple_ensemble/test.sh [hunk 10]
```diff
@@ -146,6 +147,119 @@ set -e
...
+mkdir -p ${MODEL_DIR}/slow_consumer/1
+cp ../python_models/ground_truth/model.py ${MODEL_DIR}/slow_consumer/1
+cp ../python_models/ground_truth/config.pbtxt ${MODEL_DIR}/slow_consumer/
+sed -i 's/name: "ground_truth"/name: "slow_consumer"/g' ${MODEL_DIR}/slow_consumer/config.pbtxt
+
+# Create ensemble with "max_inflight_responses = 4"
+rm -rf ${MODEL_DIR}/ensemble_enabled_max_inflight_responses
+mkdir -p ${MODEL_DIR}/ensemble_enabled_max_inflight_responses/1
+cp ${MODEL_DIR}/ensemble_disabled_max_inflight_responses/config.pbtxt ${MODEL_DIR}/ensemble_enabled_max_inflight_responses/
+sed -i 's/ensemble_scheduling {/ensemble_scheduling {\n  max_inflight_responses: 4/g' \
```
Comment (yinggeh) [inline qa/L0_simple_ensemble/test.sh]: Need to show the ensemble pipeline is working when `max_inflight_responses=1.`

**Review by yinggeh — CHANGES_REQUESTED** [review]

**Review by whoisj — DISMISSED** [review]

File: docs/user_guide/decoupled_models.md (hunk 4 above)
Comment (pskiran1) [reply docs/user_guide/decoupled_models.md:98]: Thank you, you are correct. I have updated the code to apply backpressure only to the downstream steps (exclude step 0). This change ensures that we do not block the `EnsembleScheduler::Enqueue` path (request by HTTP/GRPC handler thread). And, backpressure will only be utilized in the `ResponseComplete` path.

File: qa/L0_simple_ensemble/test.sh (hunk 10 above)
Comment (pskiran1) [reply qa/L0_simple_ensemble/test.sh]: Thank you. Updated test to cover `max_inflight_requests=1`.

File: docs/user_guide/decoupled_models.md (hunk 4 above)
Comment (yinggeh) [reply docs/user_guide/decoupled_models.md:98]: I think you misunderstood. What I am trying to say is this feature not only alleviate decoupled model memory growth, it also helps the above case. I don't think it will really blocking gRPC/HTTP handler as they are normally fast. `max_inflight_requests` only throttles the slowest step. My suggestion is do what you have done here, provide a senario of unbounded memory. However, we should not only mention this particular case and should make description in https://github.com/triton-inference-server/common/pull/141/files more generic.

File: qa/L0_simple_ensemble/ensemble_backpressure_test.py [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+        """
+        Test edge case: max_inflight_requests=1.
+        """
+        self._run_inference(model_name=MODEL_ENSEMBLE_LIMIT_ONE, expected_count=16)
+
+    def test_backpressure_disabled(self):
+        """
+        Test that an ensemble model without max_inflight_requests parameter works correctly.
+        """
+        self._run_inference(model_name=MODEL_ENSEMBLE_DISABLED, expected_count=32)
```
Comment (yinggeh) [inline qa/L0_simple_ensemble/ensemble_backpressure_test.py]: Why is `expected_count` identical to MODEL_ENSEMBLE_ENABLED?

File: qa/L0_simple_ensemble/ensemble_backpressure_test.py [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+
+import numpy as np
+import test_util as tu
+import tritonclient.grpc as grpcclient
+from tritonclient.utils import InferenceServerException
+
+SERVER_URL = "localhost:8001"
+DEFAULT_RESPONSE_TIMEOUT = 60
+EXPECTED_INFER_OUTPUT = 0.5
+MODEL_ENSEMBLE_ENABLED = "ensemble_max_inflight_requests_limit_4"
```
Comment (yinggeh) [inline qa/L0_simple_ensemble/ensemble_backpressure_test.py]:
```suggestion
MODEL_ENSEMBLE_LIMIT_4 = "ensemble_max_inflight_requests_limit_4"
```

File: qa/L0_simple_ensemble/ensemble_backpressure_test.py [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+MODEL_ENSEMBLE_DISABLED = "ensemble_disabled_max_inflight_requests"
+MODEL_ENSEMBLE_LIMIT_ONE = "ensemble_max_inflight_requests_limit_1"
```
Comment (yinggeh) [inline qa/L0_simple_ensemble/ensemble_backpressure_test.py]:
```suggestion
MODEL_ENSEMBLE_LIMIT_1 = "ensemble_max_inflight_requests_limit_1"
```

File: qa/L0_simple_ensemble/ensemble_backpressure_test.py [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+                    )
+            finally:
+                triton_client.stop_stream()
+
+    def test_backpressure_limits_inflight(self):
+        """
+        Test that max_inflight_requests correctly limits concurrent
+        responses.
+        """
+        self._run_inference(model_name=MODEL_ENSEMBLE_ENABLED, expected_count=32)
```
Comment (yinggeh) [inline qa/L0_simple_ensemble/ensemble_backpressure_test.py]: What makes some `expected_count` 32 and some 16?

Comment (pskiran1) [reply qa/L0_simple_ensemble/ensemble_backpressure_test.py]: It was a typo :)

File: qa/L0_simple_ensemble/ensemble_backpressure_test.py (hunk 11 above)
Comment (pskiran1) [reply qa/L0_simple_ensemble/ensemble_backpressure_test.py]: Its name might be confusing here, it is actually an input for the model, which will return responses equal to the provided input. I renamed it to `expected_responses_count` and added a default value in `_run_inference()`.

File: qa/L0_simple_ensemble/ensemble_backpressure_test.py [hunk 15]
```diff
@@ -105,14 +105,14 @@ def _collect_responses(self, user_data):
 
         return responses
 
-    def _run_inference(self, model_name, expected_count):
+    def _run_inference(self, model_name, expected_reposnes_count=32):
```
Comment (yinggeh) [inline qa/L0_simple_ensemble/ensemble_backpressure_test.py]:
```suggestion
    def _run_inference(self, model_name, expected_responses_count=32):
```

File: docs/user_guide/decoupled_models.md (hunk 4 above)
Comment (pskiran1) [reply docs/user_guide/decoupled_models.md:98]: Updated the documentation and description in the config proto to be more generic, emphasizing that the decoupled model is one of the use cases. Please let me know if any further improvements are needed in the documentation.

File: docs/user_guide/ensemble_models.md [hunk 16]
```diff
@@ -183,17 +183,17 @@ performance, you can use
...
+An *inflight request* refers to an intermediate request generated by an upstream model that is held in memory until it is processed by a downstream model within an ensemble pipeline. When upstream models process requests significantly faster than downstream models, these in-flight requests can accumulate and potentially lead to unbounded memory growth. his problem occurs when there is a speed mismatch between different steps in the pipeline and is particularly common in *decoupled models* that produce multiple responses per request more quickly than downstream models can process them.
```
Comment (yinggeh) [inline docs/user_guide/ensemble_models.md]:
```suggestion
An *inflight request* refers to an intermediate request generated by an upstream model that is queued and held in memory until it is processed by a downstream model within an ensemble pipeline. When upstream models process requests significantly faster than downstream models, these in-flight requests can accumulate and potentially lead to unbounded memory growth. This problem occurs when there is a speed mismatch between different steps in the pipeline and is particularly common in *decoupled models* that produce multiple responses per request more quickly than downstream models can consume.
```

File: docs/user_guide/ensemble_models.md
Comment (yinggeh) [inline docs/user_guide/ensemble_models.md:243]: Add a note here say that this delay will not cause intermediate inflight requests cancel/timeout internally, but the client should be aware of latency being added, etc.

File: docs/user_guide/ensemble_models.md [hunk 17]
```diff
@@ -224,18 +224,18 @@ ensemble_scheduling {
 
 ### When to Use This Feature
 
-Use `max_inflight_requests` when your ensemble includes:
-* **Decoupled models** that produce multiple responses per request
-* **Speed mismatch**: Upstream models significantly faster than downstream models
-* **Memory constraints**: Limited GPU/CPU memory available
+Use `max_inflight_requests` when your ensemble pipeline includes:
+* **Processing bottlenecks**: When upstream steps (e.g., preprocessing) are significantly faster than downstream steps (e.g., inference).
```
Comment (yinggeh) [inline docs/user_guide/ensemble_models.md]:
```suggestion
```

File: docs/user_guide/decoupled_models.md [hunk 18]
```diff
@@ -97,9 +97,11 @@ each time with a new response. You can take a look at [grpc_server.cc](https://g
 
 ### Using Decoupled Models in Ensembles
 
-When using decoupled models within an [ensemble](ensemble_models.md), you may encounter unbounded memory growth if a decoupled model produces responses faster than downstream models can consume them, because these responses are buffered until the downstream models are ready to process them. To address this, Triton provides the `max_inflight_requests` configuration field, which limits the number of concurrent inflight requests at each ensemble step.
+When using decoupled models within an [ensemble pipeline](ensemble_models.md), you may encounter unbounded memory growth if the decoupled model produces responses faster than downstream models can consume them.
 
-For more details and examples, see [Managing Memory Usage in Ensembles with Decoupled Models](ensemble_models.md#managing-memory-usage-in-ensembles-with-decoupled-models).
+To prevent this, use the `max_inflight_requests` configuration field. This field sets the limit on the number of concurrent inflight requests permitted at each ensemble step for each inference request to control memory accumulation.
```
Comment (yinggeh) [inline docs/user_guide/decoupled_models.md]:
```suggestion
To prevent unbounded memory growth in this senario, consider using the `max_inflight_requests` configuration field. This field limits the maximum number of concurrent inflight requests permitted at each ensemble step for each inference request.
```

File: docs/user_guide/ensemble_models.md
Comment (yinggeh) [reply docs/user_guide/ensemble_models.md:243]: And in `model_config.proto`

Comment (pskiran1) [reply docs/user_guide/ensemble_models.md:243]: Added note in both places.
